### PR TITLE
expression: rewrite ConvertTz built-in functions

### PR DESCRIPTION
**Pull request by simon-xia**
Implement the `ConvertTz` builtin function.

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -1760,7 +1760,71 @@ type builtinConvertTzSig struct {
...
 // See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_convert-tz
 func (b *builtinConvertTzSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("CONVERT_TZ")
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	if args[0].IsNull() || args[1].IsNull() || args[2].IsNull() {
+		return d, errors.Trace(err)
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: s/return d, errors.Trace(err)/return

File: expression/builtin_time.go [hunk 2]
```diff
@@ -1760,7 +1760,71 @@ type builtinConvertTzSig struct {
...
+	}
+
+	sc := b.ctx.GetSessionVars().StmtCtx
+	date, err := convertToTime(sc, args[0], mysql.TypeDatetime)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	if date.IsNull() {
+		return d, nil
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: return

File: expression/builtin_time.go [hunk 3]
```diff
@@ -1760,7 +1760,71 @@ type builtinConvertTzSig struct {
...
+		if err != nil {
+			return d, errors.Trace(err)
+		}
+
+		d.SetMysqlTime(types.Time{
+			Time: types.FromGoTime(t.In(ttz)),
+			Type: mysql.TypeDatetime,
+			Fsp:  0,
+		})
+		return d, nil
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: return

File: expression/builtin_time.go [hunk 4]
```diff
@@ -1760,7 +1760,71 @@ type builtinConvertTzSig struct {
...
+		}
+
+		d.SetMysqlTime(types.Time{
+			Time: types.FromGoTime(t.Add(timeZone2Hour(toTZ) - timeZone2Hour(fromTZ))),
+			Type: mysql.TypeDatetime,
+			Fsp:  0,
+		})
+	}
+
+	return d, nil
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: return

File: expression/builtin_time.go [hunk 5]
```diff
@@ -1760,7 +1760,71 @@ type builtinConvertTzSig struct {
...
+
+	fromTZ := args[1].GetString()
+	toTZ := args[2].GetString()
+
+	r, _ := regexp.Compile(`(\+|-)([0-9]|0[0-9]|1[0-3]):00`)
+	fmatch := r.MatchString(fromTZ)
+	tmatch := r.MatchString(toTZ)
+
+	if !fmatch && !tmatch {
+
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: remove this line

File: expression/builtin_time_test.go [hunk 6]
```diff
@@ -1426,3 +1426,37 @@ func (s *testEvaluatorSuite) TestSecToTime(c *C) {
...
+	}{
+		{"2004-01-01 12:00:00", "GMT", "MET", "2004-01-01 13:00:00"},
+		{"2004-01-01 12:00:00", "-01:00", "-12:00", "2004-01-01 01:00:00"},
+		{"2004-01-01 12:00:00", "+00:00", "-11:00", "2004-01-01 01:00:00"},
+		{"2004-01-01 12:00:00", "+00:00", "-08:00", "2004-01-01 04:00:00"},
+		{"2004-01-01 12:00:00", "+00:00", "+10:00", "2004-01-01 22:00:00"},
+		{"2004-01-01 12:00:00", "+00:00", "+11:00", "2004-01-01 23:00:00"},
+		{"2004-01-01 12:00:00", "+00:00", "+12:00", "2004-01-02 00:00:00"},
+		{"2004-01-01 18:00:00", "+00:00", "+12:00", "2004-01-02 06:00:00"},
+		{"2004-01-01 12:00:00", "-00:00", "+13:00", "2004-01-02 01:00:00"},
```
Comment (XuHuaiyu) [inline expression/builtin_time_test.go]: add test case:
SELECT CONVERT_TZ('2004-01-01 12:00:00.111','+00:00','+12:39'); ==> "2004-01-02 00:39:00.111"

and add some invalid test cases, such as:
SELECT CONVERT_TZ('2004-01-01 12:00:00','+00:00','-13:00'); ==> NULL

File: util/types/time.go [hunk 7]
```diff
@@ -2275,6 +2275,7 @@ func parseOrdinalNumbers(input string) (value int, remain string) {
 	return -1, input
 }
 
+//DateFSP get fsp from date string
```
Comment (XuHuaiyu) [inline util/types/time.go]: 1. s/get/gets
2. add `.` at the end of the comment.

File: expression/builtin_time.go [hunk 8]
```diff
@@ -1762,69 +1766,75 @@ type builtinConvertTzSig struct {
 func (b *builtinConvertTzSig) eval(row []types.Datum) (d types.Datum, err error) {
 	args, err := b.evalArgs(row)
 	if err != nil {
-		return d, errors.Trace(err)
+		return
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: if err != nil, please use return d, errors.Trace(err) for tracing the error in the log.
the same as line 1784, 1803, 1808, 1811

File: expression/builtin_time.go [hunk 9]
```diff
@@ -1762,69 +1766,75 @@ type builtinConvertTzSig struct {
...
 	tmatch := r.MatchString(toTZ)
 
 	if !fmatch && !tmatch {
-
-		ftz, err := time.LoadLocation(fromTZ)
-		if err != nil {
-			return d, errors.Trace(err)
+		ftz, er := time.LoadLocation(fromTZ)
+		if er != nil {
+			return
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time.go [hunk 10]
```diff
@@ -1762,69 +1766,75 @@ type builtinConvertTzSig struct {
...
+	r, _ := regexp.Compile(`(^(\+|-)(0?[0-9]|1[0-2]):[0-5]?\d$)|(^\+13:00$)`)
 	fmatch := r.MatchString(fromTZ)
 	tmatch := r.MatchString(toTZ)
 
 	if !fmatch && !tmatch {
-
-		ftz, err := time.LoadLocation(fromTZ)
-		if err != nil {
-			return d, errors.Trace(err)
+		ftz, er := time.LoadLocation(fromTZ)
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: s/er/err
the same as line 1806, 1811, 1825

File: expression/builtin_time.go [hunk 11]
```diff
@@ -162,9 +162,9 @@ func convertTimeToMysqlTime(t time.Time, fsp int) (types.Time, error) {
 	}, nil
 }
 
-func convertToTime(sc *variable.StatementContext, arg types.Datum, tp byte) (d types.Datum, err error) {
+func convertToTimeWithFsp(sc *variable.StatementContext, arg types.Datum, tp byte, fsp int) (d types.Datum, err error) {
 	f := types.NewFieldType(tp)
-	f.Decimal = types.MaxFsp
+	f.Decimal = fsp
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: May we check it here?
``` go
if fsp > types.MaxFsp {
    fsp = types.MaxFsp
}
```

File: expression/util.go [hunk 12]
```diff
@@ -264,6 +267,19 @@ func ConvertCol2CorCol(cond Expression, corCols []*CorrelatedColumn, outerSchema
 	return cond
 }
 
+// tz has format `(^(\+|-)(0?[0-9]|1[0-2]):[0-5]?\d$)|(^\+13:00$)`
+func timeZone2Duration(tz string) time.Duration {
```
Comment (XuHuaiyu) [inline expression/util.go]: comment for this function.

File: expression/builtin_time.go [hunk 13]
```diff
@@ -1760,7 +1764,77 @@ type builtinConvertTzSig struct {
...
+	date, err := convertToTimeWithFsp(sc, args[0], mysql.TypeDatetime, fsp)
+	if err != nil {
+		return
+	}
+
+	if date.IsNull() {
+		return
+	}
+
+	mysqlTime := date.GetMysqlTime()
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: `mysqlTime` is ambiguous,
can we rename `date` to `arg0`, rename `mysqlTime` to `dt` which is used in MySQL manual?

**Comment by shenli** [comment]
@XuHuaiyu PTAL

File: expression/util.go [hunk 14]
```diff
@@ -264,6 +267,20 @@ func ConvertCol2CorCol(cond Expression, corCols []*CorrelatedColumn, outerSchema
 	return cond
 }
 
+// timeZone2Duration convert CONVERT_TZ arg such as `+08:34` to time.Duration
```
Comment (XuHuaiyu) [inline expression/util.go]: timeZone2Duration converts timezone whose format should satisfy the regular condition \`(^(\+|-)(0?[0-9]|1[0-2]):[0-5]?\d$)|(^\+13:00$)\` to time.Duration.

**Comment by XuHuaiyu** [comment]
rest LGTM.
PTAL @shenli @tiancaiamao @hanfei1991

File: expression/util.go (hunk 14 above)
Comment (simon-xia) [inline expression/util.go]: it's ok

**Comment by ngaut** [comment]
PTAL

File: expression/builtin_time.go [hunk 15]
```diff
@@ -1760,7 +1768,74 @@ type builtinConvertTzSig struct {
...
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	if args[0].IsNull() || args[1].IsNull() || args[2].IsNull() {
+		return
+	}
+
+	sc := b.ctx.GetSessionVars().StmtCtx
+	fsp := types.DateFSP(args[0].GetString())
```
Comment (shenli) [inline expression/builtin_time.go]: MySQL also accept the first argument as an integer, not only a string.
mysql> SELECT CONVERT_TZ(20040101,'+00:00','+10:00');
+----------------------------------------+
| CONVERT_TZ(20040101,'+00:00','+10:00') |
+----------------------------------------+
| 2004-01-01 10:00:00                    |
+----------------------------------------+
1 row in set (0.00 sec)

mysql>

File: expression/builtin_time_test.go [hunk 16]
```diff
@@ -1426,3 +1426,43 @@ func (s *testEvaluatorSuite) TestSecToTime(c *C) {
 		c.Assert(result, Equals, test.expect)
 	}
 }
+
+func (s *testEvaluatorSuite) TestConvertTz(c *C) {
+	tests := []struct {
+		t       string
```
Comment (shenli) [inline expression/builtin_time_test.go]: Please add a test case that t is an integer.

Comment (shenli) [inline expression/builtin_time_test.go]: And also a test case that any argument is null.

**Comment by simon-xia** [comment]
PTAL @shenli

File: expression/builtin_time.go [hunk 17]
```diff
@@ -1797,7 +1805,78 @@ type builtinConvertTzSig struct {
...
+	if arg0.IsNull() {
+		return
+	}
+
+	dt := arg0.GetMysqlTime()
+
+	fromTZ := args[1].GetString()
+	toTZ := args[2].GetString()
+
+	r, _ := regexp.Compile(`(^(\+|-)(0?[0-9]|1[0-2]):[0-5]?\d$)|(^\+13:00$)`)
```
Comment (shenli) [inline expression/builtin_time.go]: Make the regexp string as a const and give it a meanful name.

Comment (shenli) [inline expression/builtin_time.go]: Or you could add comment for it.

File: expression/builtin_time_test.go [hunk 18]
```diff
@@ -1427,6 +1427,48 @@ func (s *testEvaluatorSuite) TestSecToTime(c *C) {
 	}
 }
 
+func (s *testEvaluatorSuite) TestConvertTz(c *C) {
+	tests := []struct {
```
Comment (shenli) [inline expression/builtin_time_test.go]: Any test case covers ! fmatch && tmatch  or fmatch && !tmatch?

Comment (simon-xia) [inline expression/builtin_time_test.go]: case `{"2004-01-01 12:00:00", "GMT", "+00:00", true, ""}` covers `! fmatch && tmatch`, and I add another case to cover `fmatch && !tmatch?`

**Comment by shenli** [comment]
LGTM
Thanks! @simon-xia

**Comment by shenli** [comment]
@XuHuaiyu PTAL

File: expression/builtin_time_test.go [hunk 19]
```diff
@@ -1427,6 +1427,49 @@ func (s *testEvaluatorSuite) TestSecToTime(c *C) {
...
+		{"2004-01-01 12:00:00", "GMT", "MET", true, "2004-01-01 13:00:00"},
+		{"2004-01-01 12:00:00", "-01:00", "-12:00", true, "2004-01-01 01:00:00"},
+		{"2004-01-01 12:00:00", "-00:00", "+13:00", true, "2004-01-02 01:00:00"},
+		{"2004-01-01 12:00:00", "-00:00", "-13:00", true, ""},
+		{"2004-01-01 12:00:00", "-00:00", "-12:88", true, ""},
+		{"2004-01-01 12:00:00", "+10:82", "GMT", false, ""},
+		{"2004-01-01 12:00:00", "+00:00", "GMT", true, ""},
+		{"2004-01-01 12:00:00", "GMT", "+00:00", true, ""},
+		{20040101, "+00:00", "+10:32", true, "2004-01-01 10:32:00"},
+		{3.14159, "+00:00", "+10:32", false, ""},
```
Comment (XuHuaiyu) [inline expression/builtin_time_test.go]: @simon-xia 
I'm not sure whether the following 2 unsual cases will keep consistent with MySQL?
mysql> select convert_tz("20040101.111", "+00:00", "+10:32");
+------------------------------------------------+
| convert_tz("20040101.111", "+00:00", "+10:32") |
+------------------------------------------------+
| 2004-01-01 21:33:00                            |
+------------------------------------------------+
1 row in set (0.00 sec)

mysql> select convert_tz(20040101.111, "+00:00", "+10:32");
+----------------------------------------------+
| convert_tz(20040101.111, "+00:00", "+10:32") |
+----------------------------------------------+
| 2004-01-01 10:32:00.000                      |
+----------------------------------------------+
1 row in set, 1 warning (0.00 sec)

Comment (simon-xia) [inline expression/builtin_time_test.go]: It seems `select convert_tz("20040101.111", "+00:00", "+10:32");` in mysql got the wrong answer

**Pull request merged**